### PR TITLE
Add retrieve for new map/multimap

### PR DESCRIPTION
**Pull request by PointKernel**
This PR introduces `retrieve` APIs for map and multimap, while also removing the outdated and odd `pair_count/retrieve` tests.

### CODE REVIEW

File: include/cuco/detail/static_map/static_map.inl [hunk 1]
```diff
@@ -612,6 +612,26 @@ static_map<Key, T, Extent, Scope, KeyEqual, ProbingScheme, Allocator, Storage>::
...
+          class ProbingScheme,
+          class Allocator,
+          class Storage>
+template <typename InputIt, typename OutputIt1, typename OutputIt2>
+std::pair<OutputIt1, OutputIt2>
+static_map<Key, T, Extent, Scope, KeyEqual, ProbingScheme, Allocator, Storage>::retrieve(
+  InputIt first,
+  InputIt last,
+  OutputIt1 output_probe,
+  OutputIt2 output_match,
```
Comment (sleeepyjack) [inline include/cuco/detail/static_map/static_map.inl]:
```suggestion
  OutputProbeIt output_probe,
  OutputMatchIt output_match,
```

Just a suggestion. Probably not worth the refactoring effort.

File: include/cuco/detail/static_map/static_map_ref.inl [hunk 2]
```diff
@@ -1420,5 +1420,73 @@ class operator_impl<
...
+            class InputProbeIt,
+            class OutputProbeIt,
+            class OutputMatchIt,
+            class AtomicCounter>
+  __device__ void retrieve(cooperative_groups::thread_block const& block,
+                           InputProbeIt input_probe_begin,
+                           InputProbeIt input_probe_end,
+                           OutputProbeIt output_probe,
+                           OutputMatchIt output_match,
+                           AtomicCounter* atomic_counter) const
```
Comment (sleeepyjack) [inline include/cuco/detail/static_map/static_map_ref.inl:1484]:
```suggestion
                           AtomicCounter& atomic_counter) const
```

I still think passing by mutable reference is the better way than passing by pointer. Anyway, we can move this to a cleanup PR.

File: include/cuco/static_multimap.cuh [hunk 3]
```diff
@@ -666,6 +666,40 @@ class static_multimap {
...
+   * the range
+   * `[first, last)`
+   *
+   * If key `k = *(first + i)` has a match `m` in the multimap, copies a `cuco::pair{k, m}` to
+   * unspecified locations in `[output_begin, output_end)`. Else, does nothing.
+   *
+   * @note This function synchronizes the given stream.
+   * @note Behavior is undefined if the size of the output range exceeds
+   * `std::distance(output_begin, output_end)`.
+   * @note Behavior is undefined if the given key has multiple matches in the set.
```
Comment (sleeepyjack) [inline include/cuco/static_multimap.cuh:680]: Undefined or rather "will only retrieve the first occurrence found"?

**Review by sleeepyjack — APPROVED** [review]
Some minor suggestions otherwise LGTM. Thanks!

File: include/cuco/detail/static_map/static_map.inl (hunk 1 above)
Comment (PointKernel) [reply include/cuco/detail/static_map/static_map.inl]: Fixed

File: include/cuco/static_multimap.cuh (hunk 3 above)
Comment (PointKernel) [reply include/cuco/static_multimap.cuh:680]: The undefined part is that if multiple matches exist in the map, the "first" occurrence is not deterministic.

**Pull request merged**